tests/suites/helpers.function

```diff
@@ -519,3 +525,51 @@ int rnd_pseudo_rand( void *rng_state, unsigned char *output, size_t len )
     return( 0 );
 }
 
+/** Mutex usage verification framework.
+ *
+ */
+
+#if defined(MBEDTLS_TEST_MUTEX_USAGE)
+typedef struct
+{
+    void (*init)( mbedtls_threading_mutex_t * );
+    void (*free)( mbedtls_threading_mutex_t * );
+    int (*lock)( mbedtls_threading_mutex_t * );
+    int (*unlock)( mbedtls_threading_mutex_t * );
+} mutex_functions_t;
+static mutex_functions_t mutex_functions;
+
+static void mbedtls_test_wrap_mutex_init( mbedtls_threading_mutex_t *mutex )
+{
+    mutex_functions.init( mutex );
+}
+
+static void mbedtls_test_wrap_mutex_free( mbedtls_threading_mutex_t *mutex )
+{
+    mutex_functions.free( mutex );
+}
+
+static int mbedtls_test_wrap_mutex_lock( mbedtls_threading_mutex_t *mutex )
+{
+    int ret = mutex_functions.lock( mutex );
+    return( ret );
+}
+
+static int mbedtls_test_wrap_mutex_unlock( mbedtls_threading_mutex_t *mutex )
+{
+    return( mutex_functions.unlock( mutex ) );
+}
+
+static void mbedtls_test_mutex_usage_init( void )
+{
+    mutex_functions.init = mbedtls_mutex_init;
+    mutex_functions.free = mbedtls_mutex_free;
+    mutex_functions.lock = mbedtls_mutex_lock;
+    mutex_functions.unlock = mbedtls_mutex_unlock;
+    mbedtls_mutex_init = &mbedtls_test_wrap_mutex_init;
```

Comment:
Using `ALT` was my first idea, but that requires either significant restructuring of `threading.c` (definitely too much for LTS branches and I'm not even keen on it for development) or arranging to compile tests (and, on development, sample programs) with a `threading_alt.h` which is a pain to set up in the build scripts.

Reply:
I think there's a misunderstanding here: I was not suggesting use of any compile-time `_ALT` thing, just calling the `set_alt()` function that's available as part of the normal `threading.h`. Mainly because using a setter function when it's available seems better style than setting the variables directly.

However, looking at the code I can see that the setter function also calls init on the global mutexes we have, which might not be desirable here. In that case, perhaps adding a comment explaining why we're not calling the setter function would be useful.

Reply:
`mbedtls_threading_set_alt` isn't compiled unless `MBEDTLS_THREADING_ALT` is defined. I wanted to minimize changes to the library, and this way I needed zero changes.

Reply:
Oh, sorry I failed to notice it was only compiled when `MBEDTLS_THREADING_ALT` is defined. Then of course it makes send to not use it.